### PR TITLE
Adds estimatecollateral RPC

### PR DESCRIPTION
**Pull request by Jouzo**
## What kind of PR is this?:

/kind feature

## What this PR does / why we need it:

Adds estimatecollateral RPC.
This RPC calculates the amount of collateral tokens a vault should have in order to take an amount of loan token and match a target collateral ratio.

## Which issue(s) does this PR fixes?:

Fixes part of #796

### CODE REVIEW

File: src/masternodes/rpc_vault.cpp [hunk 1]
```diff
@@ -1146,6 +1146,105 @@ UniValue listauctionhistory(const JSONRPCRequest& request) {
...
+        }
+
+        auto priceFeed = pcustomcsview->GetFixedIntervalPrice(loanToken->fixedIntervalPriceId);
+        if (!priceFeed.ok) {
+            throw JSONRPCError(RPC_DATABASE_ERROR, priceFeed.msg);
+        }
+
+       auto price = priceFeed.val->priceRecord[0];
+        if (!priceFeed.val->isLive(pcustomcsview->GetPriceDeviation())) {
+            throw JSONRPCError(RPC_MISC_ERROR, strprintf("No live fixed price for %s", tokenId.v));
```
Comment (bvbfan) [inline src/masternodes/rpc_vault.cpp:1202]: Same as above.

File: src/masternodes/rpc_vault.cpp [hunk 2]
```diff
@@ -1146,6 +1146,105 @@ UniValue listauctionhistory(const JSONRPCRequest& request) {
...
+        collateralSplits["DFI"] = 1;
+    }
+
+    LOCK(cs_main);
+
+    CAmount totalLoanValue{0};
+    for (const auto& [tokenId, tokenAmount] : loanAmounts.balances) {
+        auto loanToken = pcustomcsview->GetLoanTokenByID(tokenId);
+        if (!loanToken) {
+            throw JSONRPCError(RPC_DATABASE_ERROR, strprintf("(%s) is not a loan token!", tokenId));
```
Comment (bvbfan) [inline src/masternodes/rpc_vault.cpp]: loanToken->symbol is better fit? At least tokenId.v is int not string.

File: src/masternodes/rpc_vault.cpp [hunk 3]
```diff
@@ -1146,6 +1146,105 @@ UniValue listauctionhistory(const JSONRPCRequest& request) {
...
+        CAmount split = AmountFromValue(splitValue);
+
+        auto token = pcustomcsview->GetToken(tokenId);
+        if (!token) {
+            throw JSONRPCError(RPC_DATABASE_ERROR, strprintf("Token %d does not exist!", tokenId));
+        }
+
+        auto collateralToken = pcustomcsview->HasLoanCollateralToken({token->first, height});
+        if (!collateralToken || !collateralToken->factor) {
+            throw JSONRPCError(RPC_DATABASE_ERROR, strprintf("(%s) is not a valid collateral!", tokenId));
```
Comment (bvbfan) [inline src/masternodes/rpc_vault.cpp:1220]: Same.

File: src/masternodes/rpc_vault.cpp [hunk 4]
```diff
@@ -1146,6 +1146,105 @@ UniValue listauctionhistory(const JSONRPCRequest& request) {
...
+        }
+
+        auto priceFeed = pcustomcsview->GetFixedIntervalPrice(collateralToken->fixedIntervalPriceId);
+        if (!priceFeed.ok) {
+            throw JSONRPCError(RPC_DATABASE_ERROR, priceFeed.msg);
+        }
+
+        auto price = priceFeed.val->priceRecord[0];
+        if (!priceFeed.val->isLive(pcustomcsview->GetPriceDeviation())) {
+            throw JSONRPCError(RPC_MISC_ERROR, strprintf("No live fixed price for %s", tokenId));
```
Comment (bvbfan) [inline src/masternodes/rpc_vault.cpp:1230]: Same.

File: src/masternodes/rpc_vault.cpp [hunk 5]
```diff
@@ -1146,6 +1146,105 @@ UniValue listauctionhistory(const JSONRPCRequest& request) {
...
+        auto requiredValue = MultiplyAmounts(totalLoanValue, split);
+        auto collateralValue = DivideAmounts(requiredValue, price);
+        auto amountRatio = DivideAmounts(MultiplyAmounts(collateralValue, ratio), 100);
+        auto totalAmount = DivideAmounts(amountRatio, collateralToken->factor);
+
+        collateralBalances.Add({token->first, totalAmount});
+        totalSplit += split;
+    }
+    if (totalSplit != COIN) {
+        throw JSONRPCError(RPC_MISC_ERROR, strprintf("total split between collateral tokens = %d vs expected %d", totalSplit, COIN));
```
Comment (bvbfan) [inline src/masternodes/rpc_vault.cpp:1242]: Use `GetDecimaleString` for `totalSplit` and `COIN`

**Review by bvbfan — CHANGES_REQUESTED** [review]

File: src/masternodes/rpc_vault.cpp (hunk 2 above)
Comment (Jouzo) [reply src/masternodes/rpc_vault.cpp]: Cannot get `loanToken->symbol` as `!loanToken` but fixed it using `tokenId.v` as int

**Review by bvbfan — DISMISSED** [review]

**Review by bvbfan — APPROVED** [review]

**Pull request merged**